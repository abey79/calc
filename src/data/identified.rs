//! Generic type for things that have an ID, such as AST nodes.
//!
//! This type as both a generic type for the "kind" (the content of the structure) and the ID. This
//! ensure that NodeIds are not used to look up, e.g., tokens.

use std::fmt;
use std::fmt::{Debug, Display};
use std::sync::atomic::{AtomicUsize, Ordering};

#[derive(Clone, Copy, PartialEq, Default)]
pub struct Identified<K, I> {
    pub kind: K,
    pub id: I,
}

impl<K: Display, I> Display for Identified<K, I> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.kind.fmt(f)
    }
}

impl<K: Debug, I: Display> Debug for Identified<K, I> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "[{}] ", self.id)?;
        self.kind.fmt(f)
    }
}

impl<K, I> AsRef<K> for Identified<K, I> {
    fn as_ref(&self) -> &K {
        &self.kind
    }
}

pub fn new_id() -> usize {
    static COUNTER: AtomicUsize = AtomicUsize::new(0);
    COUNTER.fetch_add(1, Ordering::Relaxed)
}
