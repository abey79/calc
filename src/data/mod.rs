pub mod ast;
pub mod identified;
pub mod span;
pub mod token;
