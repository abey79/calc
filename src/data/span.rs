//! A span is a range withing a text stream.

use crate::errors::error_message::Spanned;
use std::fmt;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Loc {
    pub line: usize,
    pub col: usize,
}

impl Default for Loc {
    fn default() -> Self {
        Self { line: 1, col: 0 }
    }
}

impl Loc {
    pub fn new(line: usize, col: usize) -> Self {
        Self { line, col }
    }
}

/// A span of source code
///
/// Spans are inclusive of both start and end.
#[derive(Debug, Clone, Copy, Default)]
pub struct Span {
    pub start: Loc,
    pub end: Loc,
}

impl Span {
    pub fn new(start: Loc, end: Loc) -> Self {
        Self { start, end }
    }
}

impl fmt::Display for Span {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}", self.start.line, self.start.col)?;

        if self.start.line == self.end.line {
            write!(f, "-{}", self.end.col)
        } else {
            write!(f, "-{}:{}", self.end.line, self.end.col)
        }
    }
}

impl Spanned for Span {
    fn span(&self) -> Span {
        *self
    }
}
