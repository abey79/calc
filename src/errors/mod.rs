pub mod error_context;

use error_context::ErrorContext;
use thiserror::Error;

/// Errors generated by the tokenizer
#[derive(Error, Debug, PartialEq)]
pub enum SyntaxError {
    #[error("unexpected end of file")]
    UnexpectedEndOfFile,

    #[error("unexpected character '{0}'")]
    UnexpectedCharacter(char),
}

/// Errors generated by the tokenizer
#[derive(Error, Debug, PartialEq)]
pub enum TokenizerError {
    #[error("{1}Syntax error:{0}")]
    SyntaxError(SyntaxError, ErrorContext),
}
