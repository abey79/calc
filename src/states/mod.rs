//! This module models states (as in "state machine") between the compilation steps.
//!
//! The data associated with each state is factored in "context" structure for reusability (e.g. the
//! [`TokenStream`] context ("state data") is used in both the [`TokenizedState`] and the
//! [`ParsedState`]).

pub mod input_state;
pub mod parsed_state;
pub mod tokenized_state;

pub use input_state::*;
pub use parsed_state::*;
pub use tokenized_state::*;
