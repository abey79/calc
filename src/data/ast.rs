//! This module contains the AST data structures.
//!
//! Throughout the code, the "Node" terminology is used to refer to elements of the AST. AST nodes
//! are assigned a unique `NodeId` when created. Factory functions are also provided to make it easy
//! to build AST nodes.

use crate::data::identified::{new_id, Identified};
use crate::data::token::TokenKind;
use std::fmt;

// =================================================================================================
// DATA STRUCTURES

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct NodeId(usize);

impl NodeId {
    pub fn new() -> Self {
        Self(new_id())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnaryOpKind {
    Pos,
    Neg,
}

pub type UnaryOp = Identified<UnaryOpKind, NodeId>;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

pub type BinOp = Identified<BinOpKind, NodeId>;

pub type VarName = Identified<String, NodeId>;

#[derive(Debug, Clone, PartialEq)]
pub enum ExprKind {
    Variable(VarName),
    BinOp {
        op: BinOp,
        left: Box<Expr>,
        right: Box<Expr>,
    },
    UnaryOp {
        op: UnaryOp,
        operand: Box<Expr>,
    },
    Integer(i32),
    Float(f64),
}

pub type Expr = Identified<ExprKind, NodeId>;

#[derive(Debug, Clone, PartialEq)]
pub enum StmtKind {
    Assign { name: VarName, value: Expr },
    Print { expr: Expr },
    Expr { expr: Expr },
}

pub type Stmt = Identified<StmtKind, NodeId>;

#[derive(Debug, Clone, PartialEq)]
pub struct BlockKind {
    pub stmts: Vec<Stmt>,
}

pub type Block = Identified<BlockKind, NodeId>;

// =================================================================================================
// FACTORIES

impl From<&TokenKind> for UnaryOpKind {
    fn from(value: &TokenKind) -> Self {
        match value {
            TokenKind::Plus => Self::Pos,
            TokenKind::Minus => Self::Neg,
            _ => panic!("Invalid token kind: {:?}", value),
        }
    }
}

impl UnaryOp {
    pub fn new(kind: impl Into<UnaryOpKind>) -> Self {
        Self {
            kind: kind.into(),
            id: NodeId::new(),
        }
    }
}

impl From<&TokenKind> for BinOpKind {
    fn from(value: &TokenKind) -> Self {
        match value {
            TokenKind::Plus => Self::Add,
            TokenKind::Minus => Self::Sub,
            TokenKind::Star => Self::Mul,
            TokenKind::Slash => Self::Div,
            _ => panic!("Invalid token kind: {:?}", value),
        }
    }
}

impl BinOp {
    pub fn new(kind: impl Into<BinOpKind>) -> Self {
        Self {
            kind: kind.into(),
            id: NodeId::new(),
        }
    }
}

impl VarName {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            kind: name.into(),
            id: NodeId::new(),
        }
    }
}

impl Expr {
    pub fn variable(name: impl Into<VarName>) -> Self {
        Self {
            kind: ExprKind::Variable(name.into()),
            id: NodeId::new(),
        }
    }

    pub fn bin_op(op: BinOp, left: Expr, right: Expr) -> Self {
        Self {
            kind: ExprKind::BinOp {
                op,
                left: Box::new(left),
                right: Box::new(right),
            },
            id: NodeId::new(),
        }
    }

    pub fn unary_op(op: UnaryOp, operand: Expr) -> Self {
        Self {
            kind: ExprKind::UnaryOp {
                op,
                operand: Box::new(operand),
            },
            id: NodeId::new(),
        }
    }

    pub fn integer(value: i32) -> Self {
        Self {
            kind: ExprKind::Integer(value),
            id: NodeId::new(),
        }
    }

    pub fn float(value: f64) -> Self {
        Self {
            kind: ExprKind::Float(value),
            id: NodeId::new(),
        }
    }
}

impl Stmt {
    pub fn assign(name: impl Into<VarName>, value: Expr) -> Self {
        Self {
            kind: StmtKind::Assign {
                name: name.into(),
                value,
            },
            id: NodeId::new(),
        }
    }

    pub fn print(expr: Expr) -> Self {
        Self {
            kind: StmtKind::Print { expr },
            id: NodeId::new(),
        }
    }

    pub fn expr(expr: Expr) -> Self {
        Self {
            kind: StmtKind::Expr { expr },
            id: NodeId::new(),
        }
    }
}

impl Block {
    pub fn new(stmts: Vec<Stmt>) -> Self {
        Self {
            kind: BlockKind { stmts },
            id: NodeId::new(),
        }
    }
}

// =================================================================================================
// UTILITY TRAITS

impl fmt::Display for NodeId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl fmt::Display for BinOpKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Add => write!(f, "+"),
            Self::Sub => write!(f, "-"),
            Self::Mul => write!(f, "*"),
            Self::Div => write!(f, "/"),
        }
    }
}

impl fmt::Display for UnaryOpKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Neg => write!(f, "-"),
            Self::Pos => write!(f, "+"),
        }
    }
}

impl From<&str> for VarName {
    fn from(s: &str) -> Self {
        Self {
            kind: s.to_string(),
            id: NodeId::new(),
        }
    }
}
