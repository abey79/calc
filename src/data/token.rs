//! Tokens

use crate::data::identified::{new_id, Identified};

#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind {
    Name(String),
    Int(i32),
    Float(f64),

    // misc
    Semi,
    Assign,
    LParen,
    RParen,

    // operators
    Plus,
    Minus,
    Star,
    Slash,

    // keywords
    Print,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct TokenId(usize);

impl TokenId {
    pub fn new() -> Self {
        Self(new_id())
    }
}

pub type Token = Identified<TokenKind, TokenId>;

impl Token {
    pub fn new(kind: TokenKind) -> Self {
        Self {
            kind,
            id: TokenId::new(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct TokSpan {
    pub start: TokenId,
    pub end: TokenId,
}
