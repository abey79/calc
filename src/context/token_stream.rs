//! This context stores the tokens and their spans.

use crate::data::span::Span;
use crate::data::token::{Token, TokenId};
use crate::data::token_span::TokSpan;
use std::collections::BTreeMap;
use std::rc::Rc;

#[derive(Debug, Default)]
pub struct TokenStream {
    tokens: Vec<Rc<Token>>,
    tokens_by_id: BTreeMap<TokenId, Rc<Token>>,
    token_spans: BTreeMap<TokenId, Span>,
}

impl TokenStream {
    pub fn tokens(&self) -> &[Rc<Token>] {
        &self.tokens
    }

    pub fn token_from_id(&self, id: TokenId) -> Option<&Token> {
        self.tokens_by_id.get(&id).map(|t| &**t)
    }

    pub fn span_from_id(&self, id: TokenId) -> Option<Span> {
        self.token_spans.get(&id).copied()
    }

    pub fn push_token(&mut self, token: Token, span: Span) {
        self.token_spans.insert(token.id, span);
        let token = Rc::new(token);
        self.tokens.push(token.clone());
        self.tokens_by_id.insert(token.id, token);
    }

    pub fn span_from_tok_span(&self, tok_span: TokSpan) -> Option<Span> {
        let start = self.span_from_id(tok_span.start)?;
        let end = self.span_from_id(tok_span.end)?;
        Some(Span::new(start.start, end.end))
    }
}
