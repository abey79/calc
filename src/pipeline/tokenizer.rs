//! The tokenizer stage.
//!
//! Transforms a `RawInput` into a `TokenizedInput`.

use crate::context::token_stream::TokenStream;
use crate::data::span::{Loc, Span};
use crate::data::token::{Token, TokenKind};
use crate::errors::{Spanned, SyntaxError, TokenizerError};
use crate::states::{InputState, TokenizedState};

type Result<T> = std::result::Result<T, TokenizerError>;

pub(crate) fn tokenize(input: InputState) -> Result<TokenizedState> {
    let mut tokenizer = Tokenizer::new(input);
    tokenizer.run()?;
    Ok(TokenizedState {
        source: tokenizer.input.source,
        token_stream: tokenizer.token_stream,
    })
}

/// Implement the tokenizer stage.
struct Tokenizer {
    input: InputState,

    /// current position in the input, updated by [`next()`]
    pos: usize,

    /// current location in the input, updated by [`next()`]
    loc: Loc,

    /// starting location of the current token, updated by the lexer loop
    start_loc: Loc,

    /// current stream of token
    token_stream: TokenStream,
}

impl Tokenizer {
    fn new(input: InputState) -> Self {
        Self {
            input,
            pos: 0,
            loc: Loc::default(),
            start_loc: Loc::default(),
            token_stream: TokenStream::default(),
        }
    }

    fn source(&self) -> &str {
        self.input.as_ref()
    }

    /// Push a token into the token context.
    fn push(&mut self, kind: TokenKind) {
        let span = Span::new(self.start_loc, self.loc);
        let token = Token::new(kind, span);
        self.token_stream.push_token(token);
    }

    /// Return the next character in the input stream and update the current location.
    ///
    /// Returns `None` if the end of the input is reached.
    fn next(&mut self) -> Option<char> {
        let c = self.source().chars().nth(self.pos);
        if let Some(c) = c {
            self.pos += 1;
            if c == '\n' {
                self.loc.line += 1;
                self.loc.col = 0;
            } else {
                self.loc.col += 1;
            }
        }

        c
    }

    /// Return the next character in the input stream without updating the current location.
    ///
    /// Returns `None` if the end of the input is reached.
    fn peek(&self) -> Option<char> {
        self.source().chars().nth(self.pos)
    }

    /// Return the next character in the input stream if it matches `c` and update the current
    /// location.
    fn accept(&mut self, c: char) -> bool {
        if self.peek() == Some(c) {
            self.next();
            true
        } else {
            false
        }
    }

    fn err<T>(&self, err: SyntaxError) -> Result<T> {
        let span = Span::new(self.loc, self.loc);
        let new_err = TokenizerError::SyntaxError(err, span.to_error(&self.input.source));

        Err(new_err)
    }

    fn run(&mut self) -> Result<()> {
        while let Some(c) = self.next() {
            self.start_loc = self.loc;

            match c {
                // whitespace
                c if c.is_whitespace() => continue,
                // integer/float
                c if c.is_ascii_digit() => {
                    let mut num = c.to_string();
                    while let Some(c) = self.peek() {
                        if c.is_ascii_digit() {
                            num.push(c);
                            self.next();
                        } else {
                            break;
                        }
                    }
                    if let Some(c) = self.peek() {
                        if c == '.' {
                            num.push(c);
                            self.next();
                            while let Some(c) = self.peek() {
                                if c.is_ascii_digit() {
                                    num.push(c);
                                    self.next();
                                } else {
                                    break;
                                }
                            }
                            self.push(TokenKind::Float(num.parse().unwrap()));
                        } else {
                            self.push(TokenKind::Int(num.parse().unwrap()));
                        }
                    } else {
                        self.push(TokenKind::Int(num.parse().unwrap()));
                    }
                }
                // names/keywords
                c if c.is_ascii_alphabetic() || c == '_' => {
                    let mut name = c.to_string();
                    while let Some(c) = self.peek() {
                        if c.is_ascii_alphanumeric() || c == '_' {
                            name.push(c);
                            self.next();
                        } else {
                            break;
                        }
                    }
                    match name.as_str() {
                        // keywords
                        "print" => self.push(TokenKind::Print),
                        _ => self.push(TokenKind::Name(name)),
                    }
                }
                // misc
                ';' => self.push(TokenKind::Semi),
                '(' => self.push(TokenKind::LParen),
                ')' => self.push(TokenKind::RParen),
                ',' => self.push(TokenKind::Comma),
                '=' => self.push(TokenKind::Assign),
                '+' => self.push(TokenKind::Plus),
                '-' => self.push(TokenKind::Minus),
                '*' => self.push(TokenKind::Star),
                '/' => {
                    if self.accept('/') {
                        while let Some(c) = self.next() {
                            if c == '\n' {
                                break;
                            }
                        }
                    } else if self.accept('*') {
                        while let Some(c) = self.next() {
                            if c == '*' && self.peek() == Some('/') {
                                self.next();
                                break;
                            }
                        }
                    } else {
                        self.push(TokenKind::Slash);
                    }
                }

                c => return self.err(SyntaxError::UnexpectedCharacter(c)),
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tokenize() {
        let input = InputState::from("a = (1.3 + 3.2) * 45.1; b = a * 3.2; print 1 + 2 * 3;");
        let tokenized = tokenize(input).unwrap();

        insta::assert_debug_snapshot!(tokenized.token_stream);
    }
}
