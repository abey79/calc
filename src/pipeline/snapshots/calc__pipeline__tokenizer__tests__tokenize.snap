---
source: src/pipeline/tokenizer.rs
expression: tokenized.token_stream
---
TokenStream {
    tokens: [
        [1:1-1] Name(
            "a",
        ),
        [1:3-3] Assign,
        [1:5-5] LParen,
        [1:6-8] Float(
            1.3,
        ),
        [1:10-10] Plus,
        [1:12-14] Float(
            3.2,
        ),
        [1:15-15] RParen,
        [1:17-17] Star,
        [1:19-22] Float(
            45.1,
        ),
        [1:23-23] Semi,
        [1:25-25] Name(
            "b",
        ),
        [1:27-27] Assign,
        [1:29-29] Name(
            "a",
        ),
        [1:31-31] Star,
        [1:33-35] Float(
            3.2,
        ),
        [1:36-36] Semi,
        [1:38-42] Print,
        [1:44-44] Int(
            1,
        ),
        [1:46-46] Plus,
        [1:48-48] Int(
            2,
        ),
        [1:50-50] Star,
        [1:52-52] Int(
            3,
        ),
        [1:53-53] Semi,
    ],
}
